src/lib/transition_router/transition_router.ml

```diff
@@ -73,52 +73,76 @@ module Make (Inputs : Inputs_intf) :
     Strict_pipe.Reader.clear reader ;
     Strict_pipe.Writer.close writer
 
-  let is_transition_for_bootstrap ~frontier new_transition =
-    let root_transition =
-      Transition_frontier.root frontier
-      |> Transition_frontier.Breadcrumb.transition_with_hash |> With_hash.data
-    in
+  let is_transition_for_bootstrap root_state new_transition =
     let open External_transition in
-    let root_state = protocol_state (root_transition |> forget) in
     let new_state = protocol_state new_transition in
     Consensus.Mechanism.should_bootstrap
       ~existing:(Protocol_state.consensus_state root_state)
       ~candidate:(Protocol_state.consensus_state new_state)
 
-  let is_bootstrapping mode =
-    match mode with
-    | `Bootstrap_controller _ -> true
-    | `Transition_frontier_controller (_, _) -> false
+  let is_bootstrapping = function
+    | `Bootstrap_controller (_, _) -> true
+    | `Transition_frontier_controller (_, _, _) -> false
+
+  let get_root_state frontier =
+    Transition_frontier.root frontier
+    |> Transition_frontier.Breadcrumb.transition_with_hash |> With_hash.data
+    |> External_transition.forget |> External_transition.protocol_state
+
+  module Broadcaster = struct
+    type 'a t = {var: 'a ref; f: 'a -> unit}
```

Comment:
maybe it's more idiomatic to make this a `mutable` field than to use an explicit `ref`

Reply:
good catch 👍